libs/wingc/src/parser.rs

```diff
@@ -131,6 +132,138 @@ static RESERVED_WORDS: phf::Set<&'static str> = phf_set! {
 	"Object",
 };
 
+pub struct ParseProjectOutput {
+	pub files: Files,
+	pub tree_sitter_trees: IndexMap<PathBuf, tree_sitter::Tree>,
+	pub asts: IndexMap<PathBuf, Scope>,
+	pub topo_sorted_files: Vec<PathBuf>,
+}
+
+fn normalize_path(path: &Path, relative_to: Option<&Path>) -> PathBuf {
+	let path = if path.starts_with("./") {
+		path.strip_prefix("./").unwrap()
+	} else {
+		path
+	};
+	if path.is_absolute() {
+		path.to_path_buf()
+	} else {
+		relative_to
+			.map(|p| p.parent().unwrap_or_else(|| Path::new(".")).join(path))
+			.unwrap_or_else(|| path.to_path_buf())
+	}
+}
+
+pub fn parse_wing_project(init_path: &Path) -> ParseProjectOutput {
+	// a stack of files that need to be parsed
+	let mut needs_parsing: Vec<PathBuf> = vec![normalize_path(init_path, None)];
+	// store all source files in memory so we can look up spans
+	let mut files = Files::new();
+	// collect all tree sitter trees in case they are needed by the LSP
+	let mut tree_sitter_trees = IndexMap::new();
+	// collect all parsed trees so they can be passed to later phases of the compiler
+	let mut asts = IndexMap::new();
+	// build a dependency graph of all files
+	let mut dep_graph = DiGraph::<PathBuf, ()>::new();
+	// keep track of mapping from paths to node indices in the graph
+	let mut path_to_node: IndexMap<PathBuf, NodeIndex> = IndexMap::new();
+
+	// we can reuse the tree sitter parser
+	let language = tree_sitter_wing::language();
+	let mut tree_sitter_parser = tree_sitter::Parser::new();
+	tree_sitter_parser.set_language(language).unwrap();
+
+	while !needs_parsing.is_empty() {
+		let curr_path = needs_parsing.remove(0);
+		if files.contains_file(&curr_path) {
+			continue;
+		}
+
+		let source = match fs::read(&curr_path) {
+			Ok(source) => source,
+			Err(err) => {
+				report_diagnostic(Diagnostic {
+					message: format!("Error reading source file: {}: {:?}", curr_path.display(), err),
```

Comment:
I suspect rust io errors are `Dispaly`able.
```suggestion
					message: format!("Error reading source file: {}: {}", curr_path.display(), err),
```